pwem/protocols/protocol_set_filter.py

```diff
@@ -199,11 +225,52 @@ def formulaStep(self):
             modifiedSet.setMicrographs(inputSet.getMicrographs())
         self.createOutput(modifiedSet)
 
+    def topRankStep(self):
+        """
+        Goes through all items in the input set and takes the number/proportion of items with a higher/lower value
+        of the chosen attribute
+        """
+        inputSet = self.inputSet.get()
+        modifiedSet = inputSet.create(self._getExtraPath())
+
+        attrName = self.topRankAttribute.get().split('.')[1]
+        attributeValues = self.getAttributeValues(attrName, inputSet)
+        itemsIdx = list(range(len(attributeValues)))
+
+        if self.topRankOrder.get() == self.HIGHER:
+            attributeValues, itemsIdx = zip(*sorted(zip(attributeValues, itemsIdx), reverse=True))
+        elif self.topRankOrder.get() == self.LOWER:
+            attributeValues, itemsIdx = zip(*sorted(zip(attributeValues, itemsIdx)))
+
+        if self.topRankType.get() == self.NUMBER:
+            finalNumber = self.topRankNumber.get()
+        else:
+            finalNumber = round(self.topRankProportion.get() * len(inputSet))
+        itemsIdx = itemsIdx[:finalNumber]
+
+        added = 0
+        for i, item in enumerate(inputSet):
+            if i in itemsIdx:
+                modifiedSet.append(item.clone())
+                added += 1
+            if added == finalNumber:
+                break
+
+        self.createOutput(modifiedSet)
+
+
+    def getAttributeValues(self, attribute, iSet):
```

Comment:
this implies reading all set and putting it in a list. Should be done differently. Using the arguments in iterItems: orderby and maybe limit